### DISCUSSION TITLE
Idea for Player-Made Bounties

### DISCUSSION THREAD

**Discussion by TheSexyYeti** [ejektaflex/Bountiful#252] [Mod Ideas]
I don't know if this has been suggested. But why not make it so other players on a multiplayer server can add their own decrees to the bounty board for players to complete. They added the reward and what they want done (Farming a certain item, hunting another player, ect.). When the task is done, the player who added the decree get's a message. If the decree was a farming type task. The player who added the decree has to come back to the bounty board to collect the items. Or maybe add a mailbox for players to add to their bases to check when the task is done and collect any farmed items from.

**Comment by ejektaflex** [comment]
Yeah, I've been thinking for a while about this, and I think that it's getting to be time to add it. I'm thinking of adding a new block with new GUI with the following:

Tab A (creation):
* Bounty type picker (item or entity for now)
* Search bar (for picking objectives/rewards
* 3 objective slots and 2 reward slots, plus a button next to them (+ for add, x for clear)
* a "done" button that lights up once you've added both a valid objective and a valid reward

Creating a bounty here will make a "posting" of some sort which you can bring to a Bounty Board and activate, putting it onto the board on a second tab specifically for player made bounties.

Tab B (parcels):
* an inventory containing reward parcels. Once somebody turns in a player reward to a bounty board, a reward parcel is sent to this spot. Checking this block will show you all of your reward parcels that you've received from other players completing bounties, plus parcels generated from uncompleted bounties (if a bounty is not completed, the rewards are returned here as a parcel).

Technical notes:
* all player created bounties should be assigned a unique ID. Every time a criterion for the player-made bounty is pinged on the server side, we should check to see if that bounty is still valid (to make sure somebody else hasn't turned it in already. If so, then we need to set the bounty to say [Cancelled by Owner]. This check can additionally be made for all players on the server each time a player bounty is turned in. Heck, even each time a player is loaded, if we can check that, I suppose (do criterion checks catch this? Is that an inventory pickup? Not sure).

Things to consider:
* How should players remove their bounties from the board early, before they've expired?
* There should be an option to make player made bounties never expire
* 5 is probably a good limit for custom bounties per player, but I'm not quite sure yet. Actually, if we don't use a slot inventory for this, then we can make the number configurable and completely arbitrary.

**Comment by ejektaflex** [comment]
As for the actual bounty board, we'd need one new tab:

 Tab B (player bounties):
* You can find player bounties here, with a similar inventory to the first, normal tab
*  The Decree slots would be replaced with three buttons: scroll up (top), scroll down (bottom) and switch between player postings and other player's postings (middle). When viewing your own postings, you can opt to take them off the board to remove them for all players. Hovering over the posting shows how many times people have taken it already. If repeatable, also shows how many times it's been completed.

Admin/Repeatable Bounties
I guess we should allow admins to create bounties that don't expire, and can be repeated. This means that consequently, these bounties would not generate reward parcels either. We would benefit from a command to add objectives/rewards to the Posting they are holding based on pool ID. As such, they could in theory add other types of objectives/rewards to their custom Posting that are not normally possible, such as Command or Criterion type bounties.
E.g. `/bo posting add my_pool my_id [amount]`

Data Structures:

Both a Bounty and a Posting have very similar data structures. We can probably pull them out into a common structure, or have them both inherit from the same base class (probably easier)

Namely, both have:
* An objective list
* A reward list
* A unique UUID of the bounty. Used for player bounty checking when a posting is completed or retracted.
* Requestor (who made it). Used to inform people who created the bounty, and by Postings to check whether you can edit the posting.

Only bounties have:
* Rarity
* A time to complete (probably should never be null; changing config to make them expire would revert back to their default times with logic instead of data)

**Comment by ejektaflex** [comment]
Miscellanea:
* the Parcel management block should look like a sheet of paper drawn out and laid over a big crate. Probably with chutes or something on the side where you can take out parcels.
* the Player Bounty system is global. We don't want to worry about players having to find specific boards or getting lost or living far apart.
* Possible recipe for this new block:
  * B A B
  * P E P
  * P L P
* Where:
  * B = Bamboo
  * A = Paper
  * P = Plank
  * L = Log
  * E = Emerald
* Right clicking on another player to submit the bounty directly to them should be possible, assuming inventory contents permit it (if their inventory is full, don't give them a parcel. If your inventory is full, we don't want to spill extra things on the ground)
